File: 백준/Gold/17298. 오큰수/오큰수.java
import java.util.*;
import java.io.*;
public class Main {
    public static void main(String [] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        int [] array = new int[n];
        for(int i=0; i<n; i++){
            array[i] = Integer.parseInt(st.nextToken());
        }
        Stack<Integer> stack = new Stack<>();

        int [] answer = new int[n];
        for(int i=0; i<n; i++){
            while(!stack.isEmpty()) {
                if (array[stack.peek()] < array[i]) {
                    answer[stack.pop()] = array[i];
                }
                else{
                    break;
                }
            }
            stack.push(i);
        }

        while(!stack.isEmpty()){
            answer[stack.pop()] = -1;
        }

        for(int i=0; i<n; i++){
            bw.write(answer[i] + " ");
        }
        bw.flush();
        bw.close();
    }
}

File: 백준/Gold/17298. 오큰수/오큰수.py
# 오큰 수 구하기(골드4)
import sys
input = sys.stdin.readline
n = int(input())
array= list(map(int, input().split()))
answer=[-1]*(n)
stack=[]
for i in range(n-1, -1, -1):
    while stack and stack[-1] <= array[i]:
        stack.pop()
    if stack:
        answer[i] = stack[-1]
    stack.append(array[i])
print(*answer)